src/runtime/impure/lfortran_intrinsics.c

```diff
@@ -747,43 +747,87 @@ LFORTRAN_API void _lfortran_string_init(int size_plus_one, char *s) {
     s[size] = '\0';
 }
 
-// List  -----------------------------------------------------------------------
+// ########## List Implementation ##########
 
-struct _lcompilers_list_i32 {
+// Generic structure for lists with elements of any type
+
+struct _lcompilers_list_info {
     uint64_t n;
     uint64_t capacity;
-    int32_t *p;
+    uint64_t type_size;
+};
+
+struct _lcompilers_list {
+    struct _lcompilers_list_info* info;
+    void *p;
```

Comment:
I would put `n`, `capacity` and `type_size` directly into `_lcompilers_list`. I would not use extra pointer indirection.

Reply:
The reason I did this is because lists are shallow copied in python, so while shallow copying one list to another if you do the following,

```
dest->n = src->n;
dest->capacity = src->capacity;
dest->type_size = src->type_size;
dest->p = src->p;
```
and then call append on `dest`,
```
dest.append(1);
```
and then do,
```
print(src[0]);
```
then `src->n`, `src->capacity`, `src->type_size` would be unchanged and it would give `Out of bounds` error which would be incorrect from Python perspective. So, the question being, do we want to follow Python's rule of shallow copying lists or LPython should always deep copy lists. If you want to follow Python's rule then other than using another struct we can use array of 3 elements to store the information.

```python
Python 3.8.9 (default, Apr 13 2022, 08:48:06) 
[Clang 13.1.6 (clang-1316.0.21.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> src = []
>>> dest = src
>>> dest.append(1)
>>> src
[1] # calling dest.append updates src as well.
```

Reply:
I see. I think this is the same issue as with NumPy arrays, which we implement using deep copy, not shallow copy.

I would also deep copy lists.

Note that your approach would not work with freeing the list, as you would not know when to free it. You would have to use reference counting. But that is slow.

So the approach to this is to always deep copy everything, and in order to achieve Python semantics (which we need), we need to disallow the following case:
```
>>> src = [] # fine
>>> dest = src # fine
>>> dest.append(1) # fine
>>> src # error --- LPython and CPython would differ here, so we must disallow at compile time
```

The above (in general) can only be made with reference counting (or garbage collection), which we can implement, but then the user would need to ask explicitly, something like:
```
>>> src = rcp([]) # fine
>>> dest = rcp(src) # fine
>>> dest.append(1) # fine
>>> src # fine
```
Here `rcp` would be a no-op in CPython, but in LPython it would use reference counted pointers to track `src` and `dest`, and both can then point to the same list, and they will be freed when the reference count reaches zero.

Reply:
Well then this is not a problem anymore. Deep copy means 2 variables will diverge so doesn't matter what happens to the source variable. Just calling `memcpy` will work. Resolving this.

Reply:
> Note that your approach would not work with freeing the list, as you would not know when to free it. You would have to use reference counting. But that is slow.

Yes a one of the reasons why one cannot do parallel programming in Python because of locks being applied on reference count attribute of Python objects. So deep copy should be the best way for compiled python.

Reply:
Right. And later, we add an optimization ASR->ASR pass that rewrites this:
```
src = []
dest = src # deep copy
dest.append(1)
print(dest)
```
To this:
```
src = []
src.append(1)
print(src)
```
To avoid the extra copy.

Reply:
Note sure about removing `dest` altogether. May be someone might use `dest` and `src` as two separate lists?

Reply:
In CPython with the above syntax (i.e., no explicit call to deepcopy) you can't use `dest` and `src` as two separate lists. 

If there is explicit "copy" being done in CPython, then of course we can (and should) support that in LPython as well. That's not a problem.

Reply:
I see. So you mean we eventually will follow CPython semantics by the above removal of `dest` variable and hence removing any kind of copy altogether. Makes sense.

Reply:
I documented this at https://github.com/lcompilers/lpython/wiki/LPython-Semantics